client/js/chart/index.js

```diff
@@ -0,0 +1,131 @@
+import {SVG_NS} from '../util.js';
+
+/**
+ * Creates an SVG chart inside of `el` with the given data.
+ *
+ * @param {HTMLElement} el The container element for this chart.
+ * @param {Map<string, number[]>} rankingHistory The ranking history for each
+ * track.
+ * @param {Date[]} rankingDates The date of each history entry.
+ */
+export function createChart(el, rankingHistory, rankingDates) {
+  const svg = document.createElementNS(SVG_NS, 'svg');
+  svg.setAttribute('class', 'chart');
+  svg.append(createDefs());
+
+  const seriesContainer = document.createElementNS(SVG_NS, 'g');
+  seriesContainer.setAttribute('class', 'series-container');
+  svg.append(seriesContainer);
+
+  let index = 0;
+  const colors = ['blue', 'red', 'green', 'purple', 'orange'];
+
+  // get only the first 7 tracks for now
+  for (const history of [...rankingHistory.values()].slice(0, 7)) {
+    const color = colors[index % colors.length];
+    const series = createSeries(color, history);
+    seriesContainer.append(series);
+    index++;
+  }
+
+  el.append(svg);
+}
+
+const RUN_SCALE_X = 30;
+const RUN_SCALE_Y = 30;
+
+/**
+ * Creates a new series (set of lines on the chart for a specific song).
+ *
+ * @param {string} color The color of this series.
+ * @param {number[]} history The historical positions of this track on the
+ * leaderboard.
+ * @returns {SVGGElement} A group containing the series.
+ */
+function createSeries(color, history) {
+  const series = document.createElementNS(SVG_NS, 'g');
+  series.setAttribute('class', 'series');
+  series.style.setProperty('--run-color', color);
+  let start = 0;
+  let end = 0;
+  while (end < history.length) {
+    while (start < history.length && history[start] === null) {
+      start++;
+    }
+    end = start + 1;
+    while (end < history.length && history[end] !== null) {
+      end++;
+    }
+    series.append(...createRun(history, start, end));
+    start = end;
+  }
```

Comment:
This loop is a bit difficult to follow. Consider splitting up the calls to createRun and instead find all the run segments first, then create the runs:
```
const segments = [];
while (end < history.length) {
  // Null history nodes represent the end of a segment.
  if (history[end] !== null) {
    end++;
    continue;
  }

  if (end - start) > 0) {
    segments.push({start, end});
  }
  end++;
  start = end;
}

series.append(...segments.map(createRun));
```

Reply:
I don't really think that makes it clearer. I just added more comments. Does that work?

Reply:
It's clearer because it's a single while loop rather than 2 nested while loops. Please fix.

Reply:
Done